### PR TITLE
Implementation to enable TCP keepalive configuration.

### PR DESCRIPTION
**Pull request by shakti-das**
This is the implementation for API PR https://github.com/istio/api/pull/718. Corresponding Issue #9378. 

_Note: This PR uses a branch of the forked istio/api repo till the API PR is merged. Once the API PR is reviewed and merged, will update the vendor accordingly._

Signed-off-by: Shakti <shaktiprakash.das@salesforce.com>

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 1]
```diff
@@ -486,13 +486,54 @@ func applyConnectionPool(cluster *v2.Cluster, settings *networking.ConnectionPoo
...
 }
 
+func applyTcpKeepalive(env *model.Environment, cluster *v2.Cluster, settings *networking.ConnectionPoolSettings) {
+	var keepaliveProbes uint32
+	var keepaliveTime *types.Duration
+	var keepaliveInterval *types.Duration
+
+	upstreamConnectionOptions := &v2.UpstreamConnectionOptions{
+		TcpKeepalive: &core.TcpKeepalive{},
+	}
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: doesn't this mean tcp keepalives are enabled for all clusters irrespective of whether it was enabled globally or locally via destination rules?

Comment (shakti-das) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: That's true. This code creates empty tcp_keepalive. Just checked envoy code, has_tcp_keepalive will be true and that sets ENVOY_SOCKET_SO_KEEPALIVE. Will rectify it.

**Comment by hzxuzhonghu** [comment]
@shakti-das It is circleci, which can not be triggered from github page now. https://github.com/istio/istio/wiki/Working-with-CircleCI

**Comment by shakti-das** [comment]
I get the information message on my personal CircelCI `Resource class xlarge for docker is not available for your project. Default class medium will be used. Please contact your CSM person or our support team to whitelist your project.` Probably because of free account.

So, I tried removing personal CircelCI setup and re-run on the project's CircleCI by using the github commands. What is the correct way to get this working?

**Comment by ymesika** [comment]
@shakti-das you need to remove the project from your personal CircleCI account so the tests can run within the Istio account.

**Comment by hklai** [comment]
@rshriram @shakti-das some linter checks had failed in this PR. I suppose all outstanding PRs are now blocked. Please refrain from manual merge in case of linter failure in master and release-1.1.

Reverting this for now.

**Comment by shakti-das** [comment]
@hklai @rshriram The lint CI checks had failed before reaching pilot/pkg/networking, so didn't find the lint error before. Ran the lint test locally on the package and found errors of this type:
```
pilot/pkg/networking/core/v1alpha3/cluster_test.go:340:6: var destinationRuleTcpKeepalive should be destinationRuleTCPKeepalive (golint)
	var destinationRuleTcpKeepalive *networking.ConnectionPoolSettings_TCPSettings_TcpKeepalive
```

It's bit confusing that [istio/api guidelines](https://github.com/istio/api/blob/master/GUIDELINES.md) say to avoid using acronyms (Tcp and not TCP) mentioning https://github.com/istio/api/issues/364, whereas istio/istio asks to maintain acronyms (TCP and not Tcp). Expected behavior?

**Comment by shakti-das** [comment]
Created PR https://github.com/istio/istio/pull/10159 with the fixed lint checks.

**Comment by rshriram** [comment]
@hklai I had no way of verifying the lint behavior because circleci insisted on picking up @shakti-das 's personal account instead of ours. And since his lint container didn't have enough memory, it kept failing. anyway sorry about the merge. Need to get to the bottom of this annoying circleci behavior

**Comment by hklai** [comment]
@rshriram No worries. It was not too painful to revert.

I made some changes to the circle linter job in my merge PR, including
upgrading go to 1.11. Hopefully it will work better.

**Pull request merged**